# HG changeset patch
# User Rik <rik@octave.org>
# Date 1426703407 25200
#      Wed Mar 18 11:30:07 2015 -0700
# Node ID 2e7497ff25356d9acd24b4a286dad2a306aa235c
# Parent  7f44ced38646e28116466e22cfb370201c7270aa
image.m: Don't draw new image object twice for performance (bug #44567).

* image.m: Set up axis values before calling __go_image__.  Run 'axis tight'
after creating image object only if this is a new object.

diff --git a/scripts/image/image.m b/scripts/image/image.m
--- a/scripts/image/image.m
+++ b/scripts/image/image.m
@@ -167,24 +167,23 @@ function h = __img__ (hax, do_new, x, y,
       if (any (dx > tol) || any (dy > tol))
         warning (["image: non-linear X, Y data is ignored.  " ...
                   "IMG will be shown with linear mapping"]);
       endif
     endif
 
   endif  # ! isempty (img)
 
-  h = __go_image__ (hax, "cdata", img, "xdata", xdata, "ydata", ydata,
-                         "cdatamapping", "direct", varargin{:});
-
   if (do_new && ! ishold (hax))
     ## Set axis properties for new images
-
+    ## NOTE: Do this before calling __go_image__ so that image is not drawn
+    ##       once with default auto-scale axis limits and then a second time
+    ##       with tight axis limits.
     if (! isempty (img))
-      if (isscalar (get (hax, "children")))
+      if (isempty (get (hax, "children")))
         axis (hax, "tight");
       endif
 
       if (ndims (img) == 3)
         if (isinteger (img))
           cls = class (img);
           mn = intmin (cls);
           mx = intmax (cls);
@@ -193,16 +192,25 @@ function h = __img__ (hax, do_new, x, y,
       endif
 
     endif  # ! isempty (img)
 
     set (hax, "view", [0, 90], "ydir", "reverse", "layer", "top");
 
   endif  # do_new
 
+  h = __go_image__ (hax, "cdata", img, "xdata", xdata, "ydata", ydata,
+                         "cdatamapping", "direct", varargin{:});
+
+  if (do_new && ! ishold (hax) && ! isempty (img)
+      && isscalar (get (hax, "children")))
+    ## Re-scale axis limits for an image in a new figure or axis.
+    axis (hax, "tight");
+  endif
+
 endfunction
 
 
 %!demo
 %! clf;
 %! colormap (jet (21));
 %! img = 1 ./ hilb (11);
 %! x = y = -5:5;
